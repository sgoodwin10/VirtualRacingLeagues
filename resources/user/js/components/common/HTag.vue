<script setup lang="ts">
import { computed } from 'vue';

/**
 * Props for the HTag component
 */
interface Props {
  /**
   * Heading level (1-6) - determines which heading tag to render (h1, h2, h3, etc.)
   * @default 1
   */
  level?: 1 | 2 | 3 | 4 | 5 | 6;

  /**
   * Additional CSS classes to merge with the default classes for the heading level
   * @example "mb-4 text-primary"
   */
  additionalClasses?: string;

  /**
   * Override classes - completely replaces the default classes for the heading level
   * When provided, default classes are ignored
   * @example "text-4xl font-black uppercase"
   */
  overrideClasses?: string;
}

const props = withDefaults(defineProps<Props>(), {
  level: 1,
  additionalClasses: '',
  overrideClasses: '',
});

/**
 * Default Tailwind CSS classes for each heading level
 * Following the design system standards
 */
const defaultClasses: Record<1 | 2 | 3 | 4 | 5 | 6, string> = {
  1: 'text-3xl font-bold',
  2: 'text-2xl font-bold',
  3: 'text-xl font-semibold',
  4: 'text-lg font-semibold',
  5: 'text-base font-semibold',
  6: 'text-sm font-semibold',
};

/**
 * Component tag name based on the heading level
 */
const tag = computed(() => `h${props.level}` as 'h1' | 'h2' | 'h3' | 'h4' | 'h5' | 'h6');

/**
 * Computed classes based on override/additional class logic
 * - If overrideClasses is provided, use only those classes
 * - Otherwise, use default classes and merge with additionalClasses if provided
 */
const classes = computed(() => {
  if (props.overrideClasses) {
    return props.overrideClasses;
  }

  const baseClasses = defaultClasses[props.level];

  if (props.additionalClasses) {
    return `${baseClasses} ${props.additionalClasses}`;
  }

  return baseClasses;
});
</script>

<template>
  <component :is="tag" :class="classes">
    <slot />
  </component>
</template>
