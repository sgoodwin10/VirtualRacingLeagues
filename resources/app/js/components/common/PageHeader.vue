<script setup lang="ts">
import HTag from '@app/components/common/HTag.vue';

/**
 * Props for the PageHeader component
 */
interface Props {
  /**
   * The main heading text to display
   */
  title: string;

  /**
   * Optional descriptive text to display below the heading
   */
  description?: string;
}

defineProps<Props>();
</script>

<template>
  <div class="w-full">
    <HTag additional-classes="mb-2">{{ title }}</HTag>
    <div v-if="description" class="text-gray-600">{{ description }}</div>
  </div>
</template>
