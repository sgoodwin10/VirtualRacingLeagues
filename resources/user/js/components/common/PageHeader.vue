<script setup lang="ts">
import HTag from '@user/components/common/HTag.vue';

/**
 * Props for the PageHeader component
 */
interface Props {
  /**
   * The main heading text to display
   */
  title: string;

  /**
   * Optional descriptive text to display below the heading
   */
  description?: string;
}

defineProps<Props>();
</script>

<template>
  <div>
    <HTag additional-classes="mb-2">{{ title }}</HTag>
    <p v-if="description" class="text-gray-600">{{ description }}</p>
  </div>
</template>
