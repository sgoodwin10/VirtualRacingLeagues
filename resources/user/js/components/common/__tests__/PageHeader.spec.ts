import { describe, it, expect } from 'vitest';
import { mount } from '@vue/test-utils';
import PageHeader from '../PageHeader.vue';
import HTag from '../HTag.vue';

describe('PageHeader', () => {
  describe('Rendering', () => {
    it('renders the title correctly', () => {
      const wrapper = mount(PageHeader, {
        props: {
          title: 'Test Title',
        },
      });

      expect(wrapper.text()).toContain('Test Title');
    });

    it('renders the description when provided', () => {
      const wrapper = mount(PageHeader, {
        props: {
          title: 'Test Title',
          description: 'Test description text',
        },
      });

      expect(wrapper.text()).toContain('Test description text');
    });

    it('does not render description paragraph when description is not provided', () => {
      const wrapper = mount(PageHeader, {
        props: {
          title: 'Test Title',
        },
      });

      const descriptionElement = wrapper.find('p');
      expect(descriptionElement.exists()).toBe(false);
    });

    it('renders HTag component for the title', () => {
      const wrapper = mount(PageHeader, {
        props: {
          title: 'Test Title',
        },
        global: {
          components: {
            HTag,
          },
        },
      });

      const hTag = wrapper.findComponent(HTag);
      expect(hTag.exists()).toBe(true);
      expect(hTag.text()).toBe('Test Title');
    });
  });

  describe('Styling', () => {
    it('applies mb-2 class to HTag component', () => {
      const wrapper = mount(PageHeader, {
        props: {
          title: 'Test Title',
        },
        global: {
          components: {
            HTag,
          },
        },
      });

      const hTag = wrapper.findComponent(HTag);
      expect(hTag.props('additionalClasses')).toBe('mb-2');
    });

    it('applies text-gray-600 class to description paragraph', () => {
      const wrapper = mount(PageHeader, {
        props: {
          title: 'Test Title',
          description: 'Test description',
        },
      });

      const descriptionElement = wrapper.find('p');
      expect(descriptionElement.classes()).toContain('text-gray-600');
    });
  });

  describe('Props', () => {
    it('accepts title prop as required', () => {
      const wrapper = mount(PageHeader, {
        props: {
          title: 'Required Title',
        },
      });

      expect(wrapper.props('title')).toBe('Required Title');
    });

    it('accepts description prop as optional', () => {
      const wrapper = mount(PageHeader, {
        props: {
          title: 'Test Title',
          description: 'Optional description',
        },
      });

      expect(wrapper.props('description')).toBe('Optional description');
    });

    it('handles empty description prop', () => {
      const wrapper = mount(PageHeader, {
        props: {
          title: 'Test Title',
          description: '',
        },
      });

      // Empty description should still render the paragraph element (v-if checks for truthiness)
      const descriptionElement = wrapper.find('p');
      expect(descriptionElement.exists()).toBe(false);
    });
  });

  describe('Snapshot', () => {
    it('matches snapshot with title only', () => {
      const wrapper = mount(PageHeader, {
        props: {
          title: 'Test Title',
        },
        global: {
          components: {
            HTag,
          },
        },
      });

      expect(wrapper.html()).toMatchSnapshot();
    });

    it('matches snapshot with title and description', () => {
      const wrapper = mount(PageHeader, {
        props: {
          title: 'Test Title',
          description: 'Test description text',
        },
        global: {
          components: {
            HTag,
          },
        },
      });

      expect(wrapper.html()).toMatchSnapshot();
    });
  });
});
