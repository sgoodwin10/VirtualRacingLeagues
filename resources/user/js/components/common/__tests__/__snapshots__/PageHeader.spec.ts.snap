// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`PageHeader > Snapshot > matches snapshot with title and description 1`] = `
"<div>
  <h1 class="text-3xl font-bold mb-2">Test Title</h1>
  <p class="text-gray-600">Test description text</p>
</div>"
`;

exports[`PageHeader > Snapshot > matches snapshot with title only 1`] = `
"<div>
  <h1 class="text-3xl font-bold mb-2">Test Title</h1>
  <!--v-if-->
</div>"
`;
